pkg/tiller/server.go

```diff
@@ -87,3 +149,146 @@ func checkClientVersion(ctx context.Context) error {
+		}
+		apiserver, ok := md["k8s-server"]
+		if !ok {
+			return errors.New("Server url not found")
+		}
+		basicAuth, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(tokens[0], "Basic "))
+		if err != nil {
+			return err
+		}
+		user, password := getUserPasswordFromBasicAuth(string(basicAuth))
+		if len(user) == 0 || len(password) == 0 {
+			return errors.New("Token not found")
+		}
+		kubeConfig := &rest.Config{
+			Host:     apiserver[0],
+			Username: user,
+			Password: password,
+		}
+		caData, ok := md["k8s-certificate-authority"]
+		if !ok {
+			return errors.New("CA certificate not found")
+		}
+		caCert, err := base64.StdEncoding.DecodeString(caData[0])
+		if err != nil {
+			return err
+		}
+		kubeConfig.TLSClientConfig = rest.TLSClientConfig{
+			CAData: caCert,
+		}
+		userKubeClient, err := clientset.NewForConfig(kubeConfig)
+		if err != nil {
+			return err
+		}
+		return checkUserKubeClient(userKubeClient)
+	}
+	return errors.New("Metadata from context not found")
+}
+
+func getUserPasswordFromBasicAuth(token string) (string, string) {
+	st := strings.SplitN(token, ":", 2)
+	if len(st) == 2 {
+		return st[0], st[1]
+	}
+	return "", ""
+}
+
+func checkClientCert(ctx *context.Context) error {
+	md, ok := metadata.FromContext(*ctx)
+	if ok {
+		apiserver, ok := md["k8s-server"]
+		if !ok {
+			return errors.New("Server url not found")
+		}
+		kubeConfig := &rest.Config{
+			Host: apiserver[0],
+		}
+		caData, ok := md["k8s-certificate-authority"]
+		if !ok {
+			return errors.New("CA certificate not found")
+		}
+		caCert, err := base64.StdEncoding.DecodeString(caData[0])
+		if err != nil {
+			return err
+		}
+		clientCertificate, clientKey, err := getClientCertAndKeyFromContext(ctx)
+		if err != nil {
+			return err
+		}
+		kubeConfig.TLSClientConfig = rest.TLSClientConfig{
+			CAData:   []byte(caCert),
+			KeyData:  []byte(clientKey),
+			CertData: []byte(clientCertificate),
+		}
+		userKubeClient, err := clientset.NewForConfig(kubeConfig)
+		if err != nil {
+			return err
+		}
+		return checkUserKubeClient(userKubeClient)
+	}
+	return errors.New("Metadata from context not found")
+}
+
+func checkUserKubeClient(client *clientset.Clientset) error {
+	_, err := client.DiscoveryClient.ServerVersion()
+	if err != nil {
+		return errors.New("Certificate not authorized")
+	}
+	return nil
+}
+
+func getClientCertAndKeyFromContext(ctx *context.Context) (string, string, error) {
+	md, ok := metadata.FromContext(*ctx)
+	if ok {
+		cert, ok := md["k8s-client-certificate"]
+		if !ok {
+			return "", "", errors.New("Client certificate not found")
+		}
+		key, ok := md["k8s-client-key"]
+		if !ok {
+			return "", "", errors.New("Client key not found")
+		}
+		certData, err := base64.StdEncoding.DecodeString(cert[0])
+		if err != nil {
+			return "", "", err
+		}
+		keyData, err := base64.StdEncoding.DecodeString(key[0])
+		if err != nil {
+			return "", "", err
+		}
+		return string(certData), string(keyData), nil
```

Comment:
These codes are repeating so many times. Why not call them only once and then store those somewhere and use those for calculations.